### PR TITLE
State transfer performance improvments put block phase2

### PR DESCRIPTION
**Pull request by glevkovich**
This PR includes one main improvement - putting blocks concurrently. This improves the channel throughput (MB/s) by approx 25%. It was tested also to work correctly with block chunking.
Changes rely on the interface changes done in phase1.
In addition, I had to add an IO context pool and refactor getBlock as well (using the same contexts) to simplify the code.
PR also includes a small bug fix and refactoring.
I suggest reviewing by commits to get a better understanding.

Commits:
1. BCStateTran: intro BlockIODataPool, refactor getBlock at source, fix bug.
2. BCStateTran: perform putBlock to temporary ST concurrently.
3. BCStateTran: minor refactoring (reword comments, remove unneeded code).

### CODE REVIEW

File: bftengine/src/bcstatetransfer/BCStateTran.hpp [hunk 1]
```diff
@@ -357,21 +357,80 @@ class BCStateTran : public IStateTransfer {
...
+      return (fetchingState_ == FetchingState::NotFetching) ? ST_SRC_LOG : ST_DST_LOG;
+    }
+    const size_t maxNumElements_;
+    std::deque<BlockIOContextPtr> freeQ_;
+    std::deque<BlockIOContextPtr> allocatedQ_;
+    FetchingState& fetchingState_;
+  };
+
+  BlockIODataPool ioPool_;
+  std::deque<BlockIOContextPtr> ioContexes_;
```
Comment (toly-kournik) [inline bftengine/src/bcstatetransfer/BCStateTran.hpp:434]: ioContexts_

File: bftengine/src/bcstatetransfer/BCStateTran.hpp [hunk 2]
```diff
@@ -357,21 +357,80 @@ class BCStateTran : public IStateTransfer {
...
+
+    size_t numFreeElements() const { return freeQ_.size(); }
+    size_t numAllocatedElements() const { return allocatedQ_.size(); }
+    bool empty() { return freeQ_.empty(); };
+    bool full() { return allocatedQ_.empty(); };
+    size_t maxElements() { return maxNumElements_; };
+
+    BlockIOContextPtr alloc() {
+      if (freeQ_.empty()) {
+        throw std::runtime_error("No more free elements!");
```
Comment (toly-kournik) [inline bftengine/src/bcstatetransfer/BCStateTran.hpp:390]: what happens in this case? do we want to fail the entire replica?

Comment (glevkovich) [reply bftengine/src/bcstatetransfer/BCStateTran.hpp:390]: This should never happen, it means I have a bug in my code. The caller should be responsible to make sure there are always enough elements. It is doing so by calling processCommitResultsAsync with the right flags.
if there are no more free elements in the pool, it should call processCommitResultsAsync with breakIfNotReady=false.
In practice, it will never happen anyway. We have no more than 5-6 concurrent at peak, while the batch size (used also in the source) is much larger.

File: bftengine/src/bcstatetransfer/BCStateTran.hpp (hunk 1 above)
Comment (glevkovich) [reply bftengine/src/bcstatetransfer/BCStateTran.hpp:434]: Thanks, I'll fix that.

**Comment by toly-kournik** [comment]
I think the memory pool should be implemented as a template general utility.

File: bftengine/src/bcstatetransfer/BCStateTran.cpp [hunk 3]
```diff
@@ -2300,6 +2308,55 @@ std::string BCStateTran::logsForCollectingStatus(const uint64_t firstRequiredBlo
...
+      }
+      break;
+    }
+    ConcordAssertEQ(ctx->blockId, nextCommittedBlockId_);
+    try {
+      ConcordAssertEQ(ctx->future.get(), true);
+    } catch (const std::exception &e) {
+      LOG_FATAL(getLogger(), e.what());
+      ConcordAssert(false);
+    } catch (...) {
```
Comment (toly-kournik) [inline bftengine/src/bcstatetransfer/BCStateTran.cpp:2349]: Please don't catch unknown exceptions.

File: bftengine/src/bcstatetransfer/BCStateTran.cpp [hunk 4]
```diff
@@ -2212,7 +2218,9 @@ void BCStateTran::EnterGettingCheckpointSummariesState() {
   sourceSelector_.reset();
   metrics_.current_source_replica_.Get().Set(sourceSelector_.currentReplica());
 
+  processCommitResultsAsync(false, false);
```
Comment (toly-kournik) [inline bftengine/src/bcstatetransfer/BCStateTran.cpp:2225]: why is it needed here?

File: bftengine/src/bcstatetransfer/BCStateTran.cpp [hunk 5]
```diff
@@ -2415,55 +2481,77 @@ void BCStateTran::processData() {
...
+          "Before putBlock id " << nextRequiredBlock_ << ":" << std::boolalpha << KVLOG(lastBlock, actualBlockSize));
       if (!lastBlock) {
         putBlocksStTempDT_.pause();
         betweenPutBlocksStTempDT_.start();
-        as_->getPrevDigestFromBlock(nextRequiredBlock_,
-                                    reinterpret_cast<StateTransferDigest *>(&digestOfNextRequiredBlock));
-        nextRequiredBlock_--;
-        g.txn()->setLastRequiredBlock(nextRequiredBlock_);
+        //////////////////////////////////////////////////////////////////////////
+        // Not the last block - block are committed into ST temporary blockchain
```
Comment (toly-kournik) [inline bftengine/src/bcstatetransfer/BCStateTran.cpp:2505]: ST doesn't know anything about temporary blockchain - it's a storage implementation details

File: bftengine/src/bcstatetransfer/BCStateTran.cpp [hunk 6]
```diff
@@ -2415,55 +2481,77 @@ void BCStateTran::processData() {
...
           dst_time_between_sendFetchBlocksMsg_rec_.end();
           LOG_DEBUG(getLogger(), "Sending FetchBlocksMsg: lastInBatch is true");
           sendFetchBlocksMsg(firstRequiredBlock, nextRequiredBlock_, 0);
           break;
         }
       } else {
-        // this is the last block we need
-        // report collecting status (without vblock) into log
+        //////////////////////////////////////////////////////////////////////////
+        // This is the last block we need. After committed to ST temporary blockchain, all blocks are deleted from ST
```
Comment (toly-kournik) [inline bftengine/src/bcstatetransfer/BCStateTran.cpp:2536]: Same

File: bftengine/src/bcstatetransfer/BCStateTran.hpp [hunk 7]
```diff
@@ -425,12 +427,21 @@ class BCStateTran : public IStateTransfer {
...
+  bool oneShotTimerFlag_;  // used to control the trigger of oneShotTimer self requests
 
   // returns number of jobs pushed to queue
   uint16_t getBlocksConcurrentAsync(uint64_t nextBlockId, uint64_t firstRequiredBlock, uint16_t numBlocks);
 
+  // lastBlock: is true if we put the oldest block (firstRequiredBlock)
+  // breakIfNotReady: is true if caller would like to exit if the next future is not ready (job not ended yet).
+  //    In that case, a ONESHOT timer is invoked to check the future again soon.
+  // return: true if done procesing all futures, and false if the front one was not std::future_status::ready
+  bool processCommitResultsAsync(bool lastBlock, bool breakIfNotReady);
```
Comment (toly-kournik) [inline bftengine/src/bcstatetransfer/BCStateTran.hpp:444]: breakIfNotReady is not a clear parameter and complicates both implementation and understanding

File: bftengine/src/bcstatetransfer/BCStateTran.cpp (hunk 5 above)
Comment (glevkovich) [reply bftengine/src/bcstatetransfer/BCStateTran.cpp:2505]: OK, I'll remove some part of the comment in 2 places.

File: bftengine/src/bcstatetransfer/BCStateTran.hpp (hunk 7 above)
Comment (glevkovich) [reply bftengine/src/bcstatetransfer/BCStateTran.hpp:444]: I will change the name..
The whole idea of putting blocks concurrently relay on this flag.
If I don't break immediately, I'll wait for each block and blocks won't be put concurrently.
If you mean that the name should be renamed - I can change the name, but the flag is a must for the implementation of concurrent putBlock.
Many times, the future for the job we just invoked will not be ready, the main thread will break out and go to the handoff queue to fetch the next ItemDataMsag.

Comment (glevkovich) [reply bftengine/src/bcstatetransfer/BCStateTran.hpp:444]: new name:
waitForAllJobsToComplete
This has the opposite meaning: when true - cannot break out the loop, must wait for each future in the queue to complete.

**Comment by glevkovich** [comment]
@toly-kournik 
> In the second commit ""BCStateTran: perform putBlock to temporary ST concurrently there are no changes to related to a putBlock(). There's only a `BCStateTran::processCommitResultsAsync()` which name is misleading

But processCommitResultsAsync is not the only thing we use, the change is that we are now using the interface call putBlockAsync:
` ctx->future = as_->putBlockAsync(nextRequiredBlock_, ctx->blockData.get(), ctx->actualBlockSize, false);`
So what you've said is right for the finalize part, but the actual putBlock has changed as well - we are now putting blocks concurrently AND calling processCommitResultsAsync to finalize jobs (if possible). So I don't really understand what is wrong with the comment.
In short: put block is now concurrent by calling putBlockAsync and processCommitResultsAsync

**Review by toly-kournik — APPROVED** [review]
LGTM!
thanks Gil
